docs/frontend/frontend-npm-webpack.md

```diff
@@ -0,0 +1,116 @@
+
+To compile the code into something the browser understands we run the code through Webpack. These build/compile operations are provided as npm script to make them easy to run.
+
+We configure a webpack with a webpack encore.
+
+## How do we use them?
+
+When working with a javascripts and friends packages you should have a package.json in the root directory of your project. This declares all required dependencies to run your package (In PHP, this is similar to a composer.json file).
+
+To install all dependencies you should run `npm install` in the root directory of your project. This installs all third party dependencies in the /node_modules directory. (In PHP, this is similar to running composer). Every time you pull down new code you should ensure that all dependencies are installed or you will get errors such as Error Cannot find module 'foo' when you try to build your application. Phing target `npm` downloads packages and run build script. To add a new library, use the npm install command (for example, `npm install counterup2`).
+
+Once a dependency is installed you can use it in a JS file in your application. For example, if you install counterup2 you can import and use it:
+
+```js
+// assets/frontend/components/counterUpInit.js
+import counterUp from 'counterup2';
+
+export default function counterUpInit () {
+    document.querySelectorAll('.js-counter').forEach(counterItem => {
+        counterUp(counterItem, {
+            ...
+        })
+    }
+}
+```
+
+When compiling your application the process is clever enough to understand when a dependency has already been imported from a different file - meaning that everything is ultimately only ever imported once. However, by having to import dependencies into each file, you ensure that that particular file will work independently.
+
+If you want to add a new component that will listen to a certain event (for example), you have to import the component in the main file. For project-base, this is the `assets/frontend/app.js` file, for the administration is the` assets/admin/admin.js` file. The addition works just like a component installed over npm except that relative paths are used.
+
+```js
+    // assets/frontend/app.js
+    import './components/counterUpInit';
+    ...
+```
+
+When we are editing a javascripts and friends files, the change must go through the bundler (webpack). All javascripts and friends files are built using the `npm run build` command. But it would be impractical if we had to run a command in the console with every change. Therefore we can use `npm run watch` for development. This command checks if a file has changed and if it changes it adds the changes to the resulting bundle. The `npm run watch` command launches the webpack in development mode, which means creating source maps to help you debug your project.
```

Comment:
```suggestion
When we are editing a javascript and friends files, the change must go through the bundler (webpack). All javascript and friends files are built using the `npm run build` command. But it would be impractical if we had to run a command in the console with every change. Therefore we can use `npm run watch` for development. This command checks if a file has changed and if it changes it adds the changes to the resulting bundle. The `npm run watch` command launches the webpack in development mode, which means creating source maps to help you debug your project.
```